File: configuration-manager-python-client/configuration_manager_client/models/rules.py
"""Rules management models for Configuration Manager client."""

from dataclasses import dataclass, field
from datetime import datetime
from typing import Optional, Any, Dict, List

from .types import RuleType


@dataclass
class Rule:
    """Validation rule entity."""
    id: str
    configuration_id: str
    property_path: str
    rule_type: RuleType
    rule_config: Dict[str, Any]
    error_message: Optional[str]
    enabled: bool
    created_at: datetime
    updated_at: datetime


@dataclass
class RulesListResponse:
    """Rules list response."""
    rules: List[Rule]


@dataclass
class CreateRuleRequest:
    """Create rule request."""
    configuration_id: str
    property_path: str
    rule_type: RuleType
    rule_config: Dict[str, Any]
    error_message: Optional[str] = None
    enabled: Optional[bool] = None


@dataclass
class UpdateRuleRequest:
    """Update rule request."""
    property_path: Optional[str] = None
    rule_type: Optional[RuleType] = None
    rule_config: Optional[Dict[str, Any]] = None
    error_message: Optional[str] = None
    enabled: Optional[bool] = None


@dataclass
class RuleResponse:
    """Rule response."""
    rule: Rule


@dataclass
class ValidateRuleRequest:
    """Rule validation request."""
    configuration_id: str
    property_path: str
    value: Any


@dataclass
class RuleValidationResponse:
    """Rule validation response."""
    valid: bool
    errors: Optional[List[str]] = None
    warnings: Optional[List[str]] = None


@dataclass
class GetRulesOptions:
    """Options for getting rules."""
    configuration_id: str


@dataclass
class GetRulesForPathOptions:
    """Options for getting rules for a specific path."""
    include_inherited: Optional[bool] = None


@dataclass
class NumericRuleConfig:
    """Numeric rule configuration."""
    min: Optional[float] = None
    max: Optional[float] = None
    step: Optional[float] = None
    required: bool = False


@dataclass  
class PatternRuleConfig:
    """Pattern rule configuration."""
    pattern: str
    flags: Optional[str] = None
    required: bool = False


@dataclass
class CollectionRuleConfig:
    """Collection rule configuration."""
    allowed_values: Optional[List[Any]] = None
    min_items: Optional[int] = None
    max_items: Optional[int] = None
    unique_items: bool = False
    required: bool = False

File: configuration-manager-go-client/rules.go
package configmanager

import (
	"encoding/json"
	"time"
)

// Rule represents a validation rule entity
type Rule struct {
	ID              string          `json:"id"`
	ConfigurationID string          `json:"configurationId"`
	PropertyPath    string          `json:"propertyPath"`
	RuleType        RuleType        `json:"ruleType"`
	RuleConfig      json.RawMessage `json:"ruleConfig"`
	ErrorMessage    *string         `json:"errorMessage,omitempty"`
	Enabled         bool            `json:"enabled"`
	CreatedAt       time.Time       `json:"created_at"`
	UpdatedAt       time.Time       `json:"updated_at"`
}

// RulesListResponse represents rules list response
type RulesListResponse struct {
	Rules []Rule `json:"rules"`
}

// CreateRuleRequest represents create rule request
type CreateRuleRequest struct {
	ConfigurationID string      `json:"configurationId"`
	PropertyPath    string      `json:"propertyPath"`
	RuleType        RuleType    `json:"ruleType"`
	RuleConfig      interface{} `json:"ruleConfig"`
	ErrorMessage    *string     `json:"errorMessage,omitempty"`
	Enabled         *bool       `json:"enabled,omitempty"`
}

// UpdateRuleRequest represents update rule request
type UpdateRuleRequest struct {
	PropertyPath *string     `json:"propertyPath,omitempty"`
	RuleType     *RuleType   `json:"ruleType,omitempty"`
	RuleConfig   interface{} `json:"ruleConfig,omitempty"`
	ErrorMessage *string     `json:"errorMessage,omitempty"`
	Enabled      *bool       `json:"enabled,omitempty"`
}

// RuleResponse represents rule response
type RuleResponse struct {
	Rule Rule `json:"rule"`
}

// ValidateRuleRequest represents rule validation request
type ValidateRuleRequest struct {
	ConfigurationID string          `json:"configurationId"`
	PropertyPath    string          `json:"propertyPath"`
	Value           json.RawMessage `json:"value"`
}

// RuleValidationResponse represents rule validation response
type RuleValidationResponse struct {
	Valid    bool     `json:"valid"`
	Errors   []string `json:"errors,omitempty"`
	Warnings []string `json:"warnings,omitempty"`
}

// GetRulesOptions represents options for getting rules
type GetRulesOptions struct {
	ConfigurationID string `json:"configurationId"`
}

// GetRulesForPathOptions represents options for getting rules for a specific path
type GetRulesForPathOptions struct {
	IncludeInherited *bool `json:"includeInherited,omitempty"`
}

// NumericRuleConfig represents numeric rule configuration
type NumericRuleConfig struct {
	Min      *float64 `json:"min,omitempty"`
	Max      *float64 `json:"max,omitempty"`
	Step     *float64 `json:"step,omitempty"`
	Required bool     `json:"required"`
}

// PatternRuleConfig represents pattern rule configuration
type PatternRuleConfig struct {
	Pattern  string  `json:"pattern"`
	Flags    *string `json:"flags,omitempty"`
	Required bool    `json:"required"`
}

// CollectionRuleConfig represents collection rule configuration
type CollectionRuleConfig struct {
	AllowedValues []json.RawMessage `json:"allowedValues,omitempty"`
	MinItems      *int              `json:"minItems,omitempty"`
	MaxItems      *int              `json:"maxItems,omitempty"`
	UniqueItems   bool              `json:"uniqueItems"`
	Required      bool              `json:"required"`
}

// RulesService provides validation rules operations
type RulesService struct {
	client *HTTPClient
}

// NewRulesService creates a new rules service
func NewRulesService(client *HTTPClient) *RulesService {
	return &RulesService{client: client}
}
